### PR TITLE
[IMP] core: add typing on model and fields

### PR DESCRIPTION
**Pull request by Gorash**
[IMP] base: add typing on model and fields

----------------------------------------------------------

in order for typing to work under linux for odoo addons as well as base, you need to add symlinks (to simulate share repositories):
- to '`../../odoo/addons`' in a '`base/odoo/addons`' folder
- to '`../../addons`' in a '`community/odoo/addons`' folder
and modify the editor settings.
Example: VSCode settings.json for linter/pylance:
```
{
    "python.defaultInterpreterPath": "/usr/bin/python3",
    "emmet.extensionsPath": [
        "odoo",
        "base/odoo",
        "community/odoo",
        "enterprise/odoo",
    ],
    "python.analysis.extraPaths": [
        "odoo",
        "base/odoo",
        "community/odoo",
        "enterprise/odoo",
    ],
}
```

### CODE REVIEW

File: odoo/tools/misc.py [hunk 1]
```diff
@@ -186,7 +186,7 @@ def file_path(file_path, filter_ext=('',), env=None):
     :raise ValueError: if the file doesn't have one of the supported extensions (`filter_ext`)
     """
     root_path = os.path.abspath(config['root_path'])
-    addons_paths = odoo.addons.__path__ + [root_path]
+    addons_paths = list(odoo.addons.__path__) + [root_path]
```
Comment (kmagusiak) [inline odoo/tools/misc.py:189]: Same thing in other places.

```suggestion
    addons_paths = [*odoo.addons.__path__, root_path]
```

File: odoo/service/server.py [hunk 2]
```diff
@@ -1248,7 +1248,7 @@ def stop(self):
 server_phoenix = False
 
 def load_server_wide_modules():
-    server_wide_modules = {'base', 'web'} | set(odoo.conf.server_wide_modules)
```
Comment (kmagusiak) [inline odoo/service/server.py:1251]: You probably should leave it here. It could be another PR to remove `web` from server-wide modules.

File: odoo/modules/module.py [hunk 3]
```diff
@@ -367,6 +368,27 @@ def load_openerp_module(module_name):
         if info['post_load']:
             getattr(sys.modules[qualname], info['post_load'])()
 
+    except AttributeError as err:
+        _logger.critical("Couldn't load module %s", module_name)
+        tace = traceback.format_exc()
```
Comment (kmagusiak) [inline odoo/modules/module.py:373]: typo "tace" -> "trace"

File: odoo/addons/README.md [hunk 4]
```diff
@@ -0,0 +1,11 @@
+Addons module.
+
+This module serves to contain all Odoo addons, across all configured addons
+paths. For the code to manage those addons, see odoo.modules.
+
+Addons are made available under `odoo.addons` after
+odoo.tools.config.parse_config() is called (so that the addons paths are
+known).
+
+This module also conveniently reexports some symbols from odoo.modules.
```
Comment (kmagusiak) [inline odoo/addons/README.md:10]: Which ones?

File: odoo/fields.py [hunk 5]
```diff
@@ -25,6 +26,11 @@
 from psycopg2.extras import Json as PsycopgJson
 from difflib import get_close_matches, unified_diff
 from hashlib import sha256
+from typing import Any, TypeVar, Generic, ForwardRef, get_args, overload
+try:
```
Comment (kmagusiak) [inline odoo/fields.py:30]: For some typing information, you could import them `if typing.TYPE_CHECKING` reducing the footprint of what is loaded at runtime. You should do this especially for `Self`.

(Same in other files)

File: odoo/fields.py [hunk 6]
```diff
@@ -1711,7 +1726,7 @@ def convert_to_write(self, value, record):
         return value
 
 
-class _String(Field):
+class _String(Field[str]):
```
Comment (kmagusiak) [inline odoo/fields.py:1729]:
```suggestion
class _String(Field[str | typing.Literal[False]]):
```

same for dates and selection.

File: odoo/fields.py [hunk 7]
```diff
@@ -2967,14 +2982,14 @@ def convert_to_display_name(self, value, record):
         return value.display_name if value else False
 
 
-class _Relational(Field):
+class _Relational(Field[_T]):
```
Comment (kmagusiak) [inline odoo/fields.py:2985]:
```suggestion
class _Relational(Field[_T], Generic[_T]):
```

The class is still generic, so you need to add the `Generic` type (in the last position).
Same for other abstract types.

You also need to constrain T to a subtype of BaseModel.

File: odoo/fields.py [hunk 8]
```diff
@@ -3083,7 +3098,7 @@ class Many2one(_Relational):
     auto_join = False                   # whether joins are generated upon search
     delegate = False                    # whether self implements delegation
 
-    def __init__(self, comodel_name=Default, string=Default, **kwargs):
+    def __init__(self, comodel_name: _type[_T] = Default, string: str = Default, **kwargs):
```
Comment (kmagusiak) [inline odoo/fields.py:3101]: You need the type of the default value too. It's a little bit hard to do in python, I ended up with something like:

```python
class DefaultType(Enum): DEFAULT = "_"

Default = DefaultType.DEFAULT
```

File: odoo/fields.py [hunk 9]
```diff
@@ -44,6 +50,9 @@
 from odoo.exceptions import CacheMiss
 from odoo.osv import expression
 
+_T = TypeVar("_T", bound=Any)
+_type = type
```
Comment (kmagusiak) [inline odoo/fields.py:54]:
```suggestion
T = TypeVar("T", bound=Any)
```
Just use `type` directly and a simpler version of type var.

File: odoo/fields.py [hunk 10]
```diff
@@ -5213,7 +5228,7 @@ def write_new(self, records_commands_list):
             ])
 
 
-class Id(Field):
+class Id(Field[int]):
```
Comment (kmagusiak) [inline odoo/fields.py:5231]:
```suggestion
class Id(Field[int | NewId]):
```
... or else type the needed methods.

File: odoo/models.py [hunk 11]
```diff
@@ -592,12 +600,12 @@ class BaseModel(metaclass=MetaModel):
...
 
-    _name = None                #: the model name (in dot-notation, module namespace)
-    _description = None         #: the model's informal name
+    _name: str | None = None                #: the model name (in dot-notation, module namespace)
+    _description: str | None = None         #: the model's informal name
     _module = None              #: the model's module (in the Odoo sense)
     _custom = False             #: should be True for custom models only
 
-    _inherit = ()
+    _inherit: str | list[str] = ()
```
Comment (kmagusiak) [inline odoo/models.py:608]: You have just assigned a tuple :)

File: odoo/models.py [hunk 12]
```diff
@@ -1755,7 +1764,7 @@ def name_search(self, name='', args=None, operator='ilike', limit=100):
         return [(record.id, record.display_name) for record in records.sudo()]
 
     @api.model
-    def _name_search(self, name, domain=None, operator='ilike', limit=None, order=None):
+    def _name_search(self, name, domain=None, operator='ilike', limit=None, order=None) -> Query:
```
Comment (kmagusiak) [inline odoo/models.py:1767]: This is not true for overrides and we will remove this function soon.
You can skip name search for now.

File: odoo/models.py [hunk 13]
```diff
@@ -4349,7 +4358,7 @@ def _filter_access_rules_python(self, operation):
         dom = self.env['ir.rule']._compute_domain(self._name, operation)
         return self.sudo().filtered_domain(dom or [])
 
-    def unlink(self):
+    def unlink(self) -> True:
```
Comment (kmagusiak) [inline odoo/models.py:4361]: You need to set `Literal[True]` as a type.

File: odoo/models.py [hunk 14]
```diff
@@ -5944,7 +5953,7 @@ def _get_redirect_suggested_company(self):
     #  - the global cache is only an index to "resolve" a record 'id'.
     #
 
-    def __init__(self, env: api.Environment, ids: tuple[int | NewId], prefetch_ids: Iterable[int | NewId]):
+    def __init__(self, env: api.Environment, ids: int | NewId | tuple[int | NewId] = (), prefetch_ids: Iterable[int | NewId] = ()):
```
Comment (kmagusiak) [inline odoo/models.py:5956]:
```suggestion
    def __init__(self, env: api.Environment, ids: tuple[int | NewId] = (), prefetch_ids: Reversible[int] = ()):
```

File: odoo/models.py [hunk 15]
```diff
@@ -5983,7 +5992,7 @@ def browse(self, ids=None):
     #
 
     @property
-    def ids(self):
+    def ids(self) -> tuple[int | NewId]:
```
Comment (kmagusiak) [inline odoo/models.py:5995]:
```suggestion
    def ids(self) -> tuple[int]:
```

**Review by kmagusiak — COMMENTED** [review]
I haven't tested yet the code nor what mypy is saying, but here are the first things I have found.

**Review by Julien00859 — COMMENTED** [review]
What's about Windows? There are no symbolic links on windows...

**Comment by Gorash** [comment]
@Julien00859 no idea for Windows, and it doesn't matter. We add tools, it doesn't matter if it's not active on some OS. As long as it doesn't introduce a bug. The community that uses Windows will give us feedback that we will try to introduce.

**Review by kmagusiak — COMMENTED** [review]
Starts looking good.

mypy shows things like
```
error: Need type annotation for "child_ids"  [var-annotated] -- actually very good because if a definition of a relational field is missing a type annotation, it gives this error
error: "Partner" has no attribute "display_name"  [attr-defined]  -- because it's not declared as a field, I wonder why this is not the case
error: "_get_view_cache_key" undefined in superclass  [misc] -- more difficult as this is defined on model 'base', not sure if we can add it to the typing system
```

File: odoo/addons/base/models/res_partner.py [hunk 16]
```diff
@@ -16,6 +17,10 @@
 from odoo import api, fields, models, tools, SUPERUSER_ID, _, Command
 from odoo.exceptions import RedirectWarning, UserError, ValidationError
 
+import typing
+if typing.TYPE_CHECKING:
+    from typing_extensions import Self
```
[the remark this answers is not in the document]
Comment (Gorash) [reply odoo/addons/base/models/res_partner.py:22]: if I do this, VScode doesn't understand typing anymore. It writes "any | Self' instead of 'Self'

File: odoo/api.py [hunk 17]
```diff
@@ -34,12 +33,17 @@
...
 from odoo.tools.misc import StackMap
 
-if TYPE_CHECKING:
+import typing
+if typing.TYPE_CHECKING:
+    from collections.abc import Callable
     from odoo.sql_db import Cursor, TestCursor
     from odoo.models import BaseModel
+P = typing.ParamSpec("P")
+R = typing.TypeVar("R", bound=typing.Any)
```
[the remark this answers is not in the document]
Comment (Gorash) [reply odoo/api.py:42]: The code does not compile. For unknown attribute

File: odoo/addons/base/models/res_partner.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (kmagusiak) [reply odoo/addons/base/models/res_partner.py:22]: Ok, I think it is because you have python 3.10. What about the following (since `typing_extensions` is not installed after 3.10).

```python
try:
  from typing_extensions import Self
except ImportError:
  from typing import Self
```

File: odoo/api.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (kmagusiak) [reply odoo/api.py:42]: Ok, probably because of the `Any`. Why bind to any, could you try without  binding R?

File: odoo/api.py [hunk 18]
```diff
@@ -103,7 +107,7 @@ def propagate(method1, method2):
     return method2
 
 
-def constrains(*args):
+def constrains(*args: str) -> Callable[[Callable[P, R]], Callable[P, R]]:
```
[the remark this answers is not in the document]
Comment (Gorash) [reply odoo/api.py:110]: If I do that, VSCode does not understand the new type for constrains and depends

File: odoo/addons/base/models/res_partner.py [hunk 19]
```diff
@@ -732,7 +737,7 @@ def write(self, vals):
         return result
 
     @api.model_create_multi
```
[the remark this answers is not in the document]
Comment (Gorash) [reply odoo/addons/base/models/res_partner.py:739]: OK for `@api.model_create_multi`.
Not for `@api.model`, decorated methods can return anything.

[the remark this answers is not in the document]
Comment (kmagusiak) [reply odoo/addons/base/models/res_partner.py:743]: Yes, they can, but you can type:

```python
def model(method: T) -> T:
# or if you prefer with Callable[P, R]
```

The method can return anything, but the decorator `@api.model` returns a function of the same type it was given.

File: odoo/api.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (kmagusiak) [reply odoo/api.py:112]: We can compare on Wednesday, it is working fine for me.

File: odoo/addons/base/models/res_partner.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (Gorash) [reply odoo/addons/base/models/res_partner.py:743]: Ok, just a generic make it.
(For the previous change I think about `Callable[[T, typing.Unpack[typing.Any]], T]` usefull for create.)

File: odoo/api.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (Gorash) [reply odoo/api.py:112]: I don't have the problem anymore, it must have been due to the wrong getitem typing

File: odoo/models.py [hunk 20]
```diff
@@ -1699,7 +1713,7 @@ def _compute_display_name(self):
                 record.display_name = f"{record._name},{record.id}"
 
     @api.model
-    def name_create(self, name):
+    def name_create(self, name) -> tuple[int, str] | False:
```
Comment (kmagusiak) [inline odoo/models.py:1716]: literal

**Review by kmagusiak — APPROVED** [review]

File: odoo/fields.py [hunk 21]
```diff
@@ -38,12 +39,29 @@
...
+    try:
+        from typing_extensions import Self
+    except ImportError:
+        from typing import Self
+    from odoo.models import BaseModel
+    from collections.abc import Mapping
+    M = typing.TypeVar("M", bound=BaseModel)
+    ModelId = int | NewId
+    TypeDomain = list[str | tuple[str, str, typing.Any]]
+    TypeContext = Mapping[str, typing.Any]
```
Comment (rco-odoo) [inline odoo/fields.py:59]: Shouldn't those be defined in `models.py` and imported here?

File: odoo/models.py [hunk 22]
```diff
@@ -568,6 +577,11 @@ class BaseModel(metaclass=MetaModel):
     """
     __slots__ = ['env', '_ids', '_prefetch_ids']
 
+    env: api.Environment
+    id: int | NewId | typing.Literal[False]
```
Comment (rco-odoo) [inline odoo/models.py:581]: Maybe we should type this one as `IdType`.

File: odoo/models.py [hunk 23]
```diff
@@ -4363,7 +4377,7 @@ def _filter_access_rules_python(self, operation):
         dom = self.env['ir.rule']._compute_domain(self._name, operation)
         return self.sudo().filtered_domain(dom or [])
 
-    def unlink(self):
+    def unlink(self) -> typing.Literal[True]:
```
Comment (rco-odoo) [inline odoo/models.py:4380]: I wonder whether this one is a good idea, because
- the returned value is irrelevant (nobody checks the returned value);
- it is not guaranteed to be `True` (overrides);
- ideally it should be `None`.

File: odoo/models.py [hunk 24]
```diff
@@ -4464,7 +4478,7 @@ def unlink(self):
 
         return True
 
-    def write(self, vals):
+    def write(self, vals) -> typing.Literal[True]:
```
Comment (rco-odoo) [inline odoo/models.py:4481]:
```suggestion
    def write(self, vals: dict[str, Any]) -> typing.Literal[True]:
```

File: odoo/models.py [hunk 25]
```diff
@@ -4731,7 +4745,7 @@ def _write_multi(self, vals_list):
             parent_records._parent_store_update()
 
     @api.model_create_multi
-    def create(self, vals_list):
+    def create(self, vals_list) -> Self:
```
Comment (rco-odoo) [inline odoo/models.py:4748]:
```suggestion
    def create(self, vals_list: list[dict[str, Any]]) -> Self:
```
Maybe we should make aliases for common types:
```py
DomainType = list[str | tuple[str, str, Any]]
ContextType = dict[str, Any]
ValuesType = dict[str, Any]
```

File: odoo/models.py [hunk 26]
```diff
@@ -5710,7 +5724,7 @@ def copy_translations(self, new, excluded=()):
                     new.update_field_translations(name, translations)
 
     @api.returns('self')
-    def copy(self, default=None):
+    def copy(self, default=None) -> Self:
```
Comment (rco-odoo) [inline odoo/models.py:5727]:
```suggestion
    def copy(self, default: dict[str, Any] | None = None) -> Self:
```

File: odoo/models.py [hunk 27]
```diff
@@ -5958,7 +5972,7 @@ def _get_redirect_suggested_company(self):
     #  - the global cache is only an index to "resolve" a record 'id'.
     #
 
-    def __init__(self, env: api.Environment, ids: tuple[int | NewId], prefetch_ids: Iterable[int | NewId]):
+    def __init__(self, env: api.Environment, ids: tuple[int | NewId] = (), prefetch_ids: Reversible[int | NewId] = ()):
```
Comment (rco-odoo) [inline odoo/models.py:5975]: All three parameters are required on purpose.  Please don't change that.
```suggestion
    def __init__(self, env: api.Environment, ids: tuple[int | NewId], prefetch_ids: Reversible[int | NewId]):
```

File: odoo/models.py [hunk 28]
```diff
@@ -5969,7 +5983,7 @@ def __init__(self, env: api.Environment, ids: tuple[int | NewId], prefetch_ids:
         self._ids = ids
         self._prefetch_ids = prefetch_ids
 
-    def browse(self, ids=None):
+    def browse(self, ids=None) -> Self:
```
Comment (rco-odoo) [inline odoo/models.py:5986]:
```suggestion
    def browse(self, ids: typing.Iterable[IdType] = ()) -> Self:
```

File: odoo/models.py [hunk 29]
```diff
@@ -6023,7 +6037,7 @@ def ensure_one(self):
         except ValueError:
             raise ValueError("Expected singleton: %s" % self)
 
-    def with_env(self, env):
+    def with_env(self, env) -> Self:
```
Comment (rco-odoo) [inline odoo/models.py:6040]:
```suggestion
    def with_env(self, env: api.Environment) -> Self:
```

File: odoo/models.py [hunk 30]
```diff
@@ -6774,6 +6788,12 @@ def __repr__(self):
     def __hash__(self):
         return hash((self._name, frozenset(self._ids)))
 
+    @typing.overload
+    def __getitem__(self, key: int | typing.Sequence[slice]) -> Self: ...
```
Comment (rco-odoo) [inline odoo/models.py:6792]:
```suggestion
    def __getitem__(self, key: int | slice) -> Self: ...
```

File: odoo/modules/module.py [hunk 31]
```diff
@@ -367,6 +368,27 @@ def load_openerp_module(module_name):
...
+            raise AttributeError((
+                str(err) + "\n"
+                "To avoid circular import for the the comodel use the annotation syntax:\n"
+                "    %(field_name)s: '%(module_name)s.%(ModelClass)s' = fields.%(ttype)s(...)"
+                ) % {
+                    "field_name": field_name,
+                    "module_name": addon_name,
+                    "ModelClass": class_name,
+                    "ttype": ttype,
+                }).with_traceback(err.__traceback__) from None
```
Comment (rco-odoo) [inline odoo/modules/module.py:390]:
```suggestion
                str(err) + "\n"
                f"To avoid circular import for the the comodel use the annotation syntax:\n"
                "    {field_name}: '{module_name}.{ModelClass}' = fields.{ttype}(...)"
                )).with_traceback(err.__traceback__) from None
```

Comment (rco-odoo) [reply odoo/modules/module.py:390]: Maybe suggest to use `from __future__ import annotations`.

File: odoo/fields.py [hunk 32]
```diff
@@ -38,12 +39,29 @@
...
+import typing
+if typing.TYPE_CHECKING:
+    try:
+        from typing_extensions import Self
+    except ImportError:
+        from typing import Self
+    from odoo.models import BaseModel
+    from collections.abc import Mapping
+    M = typing.TypeVar("M", bound=BaseModel)
+    ModelId = int | NewId
```
[the remark this answers is not in the document]
Comment (kmagusiak) [reply odoo/fields.py:57]: `IdType` is used only in fields.py, but declared in models.py, it is used twice to check membership. I think it should be removed:
1. Type annotations need `|` to build union types, but `isinstance` expects a tuple.
2. Both usages are `if type(value) in IdType` should become `if type(value) is int or type(value) is NewId` (note: isinstance cannot be used because we want to ignore `False`.

File: odoo/fields.py [hunk 33]
```diff
@@ -269,11 +286,11 @@ def _read_group_many2one_field(self, records, domain, order):
         .. seealso:: :ref:`Advanced fields/Related fields <reference/fields/related>`
     """
 
-    type = None                         # type of the field (string)
+    type: str                           # type of the field (string)
     relational = False                  # whether the field is a relational one
     translate = False                   # whether the field is translated
 
-    column_type = None                  # database column type (ident, spec)
+    column_type: typing.Tuple[str, ...] | None = None  # database column type (ident, spec)
```
[the remark this answers is not in the document]
Comment (kmagusiak) [reply odoo/fields.py:293]: it should be `tuple[str, str]` (for standard collections, the constructor should be used for typing information).

File: odoo/fields.py [hunk 34]
```diff
@@ -1193,7 +1210,13 @@ def write(self, records, value):
     # Descriptor methods
     #
 
-    def __get__(self, record, owner=None):
+    @typing.overload
+    def __get__(self, record: None, owner=None) -> Self: ...
+
```
[the remark this answers is not in the document]
Comment (kmagusiak) [reply odoo/fields.py:1215]: Some very generic methods like here that have multiple usages will probably have overload information. Note that you can add a different docstring to each overloaded method to give a more specific description.

File: odoo/fields.py [hunk 35]
```diff
@@ -2967,14 +2990,14 @@ def convert_to_display_name(self, value, record):
         return value.display_name if value else False
 
 
-class _Relational(Field):
+class _Relational(Field[M], typing.Generic[M]):
```
[the remark this answers is not in the document]
Comment (kmagusiak) [reply odoo/fields.py:2993]: Before removing, I'll have to check how it will bind `M` when typechecking if `Generic` is not there.

File: odoo/fields.py (hunk 21 above)
Comment (kmagusiak) [reply odoo/fields.py:59]: I think `api.py` is a better place, that is where the environment is defined and context is related to the environment.

File: odoo/models.py (hunk 22 above)
Comment (kmagusiak) [reply odoo/models.py:581]: `IdType | typing.Literal[False]`

File: odoo/models.py (hunk 23 above)
Comment (kmagusiak) [reply odoo/models.py:4380]: Indeed, we don't need to type this now and when xmlrpc is decommissioned, we can return `None`.

File: odoo/models.py (hunk 28 above)
Comment (kmagusiak) [reply odoo/models.py:5986]: It also accepts a single `int`.

File: odoo/api.py [hunk 36]
```diff
@@ -34,12 +33,31 @@
...
+    from collections.abc import Callable
     from odoo.sql_db import Cursor, TestCursor
     from odoo.models import BaseModel
+    from .models import NewId
+
+    try:
+        from typing_extensions import Self  # noqa: F401
+    except ImportError:
+        from typing import Self  # noqa: F401
+    M = typing.TypeVar("BaseModel", bound=BaseModel)
```
Comment (kmagusiak) [inline odoo/api.py:47]: Variable name and first argument must be aligned. (same in the else block)
```suggestion
    M = typing.TypeVar("M", bound=BaseModel)
```

File: odoo/fields.py (hunk 35 above)
[the remark this answers is not in the document]
Comment (kmagusiak) [reply odoo/fields.py:2974]: After checking, you can remove `Generic` since we force users to overwrite the typing information on the field itself. You may leave it on the abstract/private field classes.

File: odoo/models.py [hunk 37]
```diff
@@ -5966,7 +5975,7 @@ def _get_redirect_suggested_company(self):
     #  - the global cache is only an index to "resolve" a record 'id'.
     #
 
-    def __init__(self, env: api.Environment, ids: tuple[int | NewId], prefetch_ids: Iterable[int | NewId]):
+    def __init__(self, env: api.Environment, ids: tuple[int | NewId], prefetch_ids: Reversible[int | NewId]):
```
Comment (kmagusiak) [inline odoo/models.py:5978]:
```suggestion
    def __init__(self, env: api.Environment, ids: tuple[int | NewId, ...], prefetch_ids: Reversible[int | NewId]):
```

File: odoo/models.py [hunk 38]
```diff
@@ -6005,7 +6014,7 @@ def browse(self, ids=None):
     #
 
     @property
-    def ids(self):
+    def ids(self) -> tuple[int]:
```
Comment (kmagusiak) [inline odoo/models.py:6017]:
```suggestion
    def ids(self) -> list[int]:
```

File: odoo/models.py [hunk 39]
```diff
@@ -64,6 +64,13 @@
 from .tools.misc import CountingStream, LastOrderedSet, ReversedIterable, unquote
 from .tools.translate import _, _lt
 
+import typing
+if typing.TYPE_CHECKING:
+    from collections.abc import Reversible
+    from .modules.registry import Registry
+from odoo.api import Self, ValuesType, IdType
```
Comment (kmagusiak) [inline odoo/models.py:71]: Used only for type checking.
```suggestion
    from odoo.api import Self, ValuesType, IdType
```

File: odoo/api.py [hunk 40]
```diff
@@ -34,12 +33,31 @@
...
+    IdType = int | NewId
```
Comment (kmagusiak) [inline odoo/api.py:48]: Probably a bug with mypy, but when you import `IdType` it is not considered valid as a type. The workaround I found is:
```suggestion
    # type written as union because not recognized by mypy when imported
    IdType = typing.Union[int, NewId]  # noqa: UP007
```

**Review by kmagusiak — APPROVED** [review]

File: odoo/modules/module.py [hunk 41]
```diff
@@ -367,6 +368,24 @@ def load_openerp_module(module_name):
...
+            addons_class = matching.group(2) or matching.group(4)
+            addon_name = addons_class.split('.')[0] if '.' in addons_class else module_name
+            class_name = addons_class.split('.')[-1]
+            raise AttributeError(
+                str(err) + "\n"
+                "To avoid circular import for the the comodel use the annotation syntax:\n"
+                f"    {field_name}: {addon_name}.{class_name} = fields.{ttype}(...)\n"
+                "and add at the beggining of the file:\n"
+                "    from __future__ import annotations"
+                ).with_traceback(err.__traceback__) from None
```
Comment (rco-odoo) [inline odoo/modules/module.py:387]: Please define a constant for that regular expression
```py
TYPED_FIELD_DEFINITION_RE = re.compile(r'''
    \b(?P<field_name>\w+)\s*
    (:\s*(?P<field_type>[^ ]*))?\s*
    =\s*fields\.(?P<field_class>One2many|Many2many|Many2one)
    (\[(?P<type_param>[^\]]+)\])?
''', re.VERBOSE)
```
and use more standard variable naming for regex matches (_matching_ is a verb)
```suggestion
        match = TYPED_FIELD_DEFINITION_RE.search(trace)
        if match and "most likely due to a circular import" in trace:
            field_name = match['field_name']
            field_class = match['field_class']
            field_type = match['field_type'] or match['type_param']
            if "." not in field_type:
                field_type = f"{module_name}.{field_type}"
            raise AttributeError(
                f"{err}\n"
                "To avoid circular import for the the comodel use the annotation syntax:\n"
                f"    {field_name}: {field_type} = fields.{field_class}(...)\n"
                "and add at the beggining of the file:\n"
                "    from __future__ import annotations"
            ).with_traceback(err.__traceback__) from None
```
Note that I've used named capturing groups, which greatly improve code readability.

File: odoo/tools/misc.py [hunk 42]
```diff
@@ -118,6 +119,14 @@
 
 NON_BREAKING_SPACE = u'\N{NO-BREAK SPACE}'
 
+
+class Sentinel(enum.Enum):
+    """Class for typing default values"""
```
Comment (rco-odoo) [inline odoo/tools/misc.py:124]:
```suggestion
class Sentinel(enum.Enum):
    """Class for typing parameters with a sentinel as a default"""
```

File: odoo/fields.py [hunk 43]
```diff
@@ -2967,14 +2971,14 @@ def convert_to_display_name(self, value, record):
...
     """ Abstract class for relational fields. """
     relational = True
-    domain = []                         # domain for searching values
-    context = {}                        # context for searching values
+    domain: DomainType = []         # domain for searching values
+    context: ContextType = {}       # context for searching values
     check_company = False
 
-    def __get__(self, records, owner=None):
+    def __get__(self, records, owner=None) -> M:
```
[the remark this answers is not in the document]
Comment (Gorash) [reply odoo/fields.py:2981]: to declare `M` (restrictive to BaseModel)

File: odoo/api.py [hunk 44]
```diff
@@ -34,12 +33,31 @@
...
+        from typing_extensions import Self  # noqa: F401
+    except ImportError:
+        from typing import Self  # noqa: F401
+    M = typing.TypeVar("M", bound=BaseModel)
+    # type written as union because not recognized by mypy when imported
+    IdType = typing.Union[int, NewId]  # noqa: UP007
+else:
+    Self = None
+    M = typing.TypeVar("M")
+    IdType = int
```
[the remark this answers is not in the document]
Comment (Gorash) [reply odoo/api.py:52]: NewId is not available in running time, can be only use for typing

[the remark this answers is not in the document]
Comment (rco-odoo) [reply odoo/api.py:52]: That's a recipe for bugs, because such a definition is **incorrect** at runtime!  Someone doing `isinstance(x, IdType)` might get a different result than expected.  And you cannot say: "Yeah, it is available, but don't use it." :wink:

File: odoo/fields.py (hunk 43 above)
[the remark this answers is not in the document]
Comment (rco-odoo) [reply odoo/fields.py:2981]: I don't see the point.  `_Relational[M]` is a subtype of `Field[M]`, and therefore `Field[M].__get__` should return a value of type `M`.

[the remark this answers is not in the document]
Comment (kmagusiak) [reply odoo/fields.py:2981]: `M` is optional too because it is already bound in the class signature.

File: odoo/api.py (hunk 44 above)
[the remark this answers is not in the document]
Comment (kmagusiak) [reply odoo/api.py:52]: Indeed, it should only be available for typing or it should be correct.
However, in normal execution, there is a lot of cycles between models.py, fields.py, api.py. Solving that cycle seems quite difficult.

[the remark this answers is not in the document]
Comment (rco-odoo) [reply odoo/api.py:52]:
```suggestion
    IdType: typing.TypeAlias = int | NewId
```
Also, solve the cyclic dependency issue by moving the definition of class `NewId` to api.py.

**Review by rco-odoo — APPROVED** [review]
@robodoo r+

**Comment by rco-odoo** [comment]
@robodoo override=ci/security

Part of the flagged code has simply been moved, and it just as fine as before.
The other case is kind of necessary (using `isinstance` wouldn't work because `isinstance(False, int)`) and the code does not leak anything in this case.

**Comment by rco-odoo** [comment]
@robodoo r+

**Comment by rco-odoo** [comment]
It's fixed now.
@robodoo r+
